### PR TITLE
Aggregations

### PR DESCRIPTION
**Pull request by gnosek**

### CODE REVIEW

**Review by marktwallace — COMMENTED** [review]
I'm still looking at files, but here are some running high level comments:

(1) These look really useful -- especially the sum_by_owner aggregates.
(2) There are a lot of new metrics here.
(3) If we  send a PR to the OSS project, we might get push-back on the overall number of metrics. There could also be multiple points of view as to whether these should be aggregated at this level -- it is good for our use case, but users that do not have the same problems of scale and data retention may think that using labels to aggregate with PromQL is better, and this would look redundant to them.
(4) Since our flag "--enable-aggregations" is orthogonal to to the "--metric-allowlist" and "--metric-denylist" flags, it would bring in aggregates for things that are denied, right? That could be a source of confusion for OSS users.

I'll keep adding comments as I go. I'm liking the code so far :-)

File: internal/store/builder.go [hunk 1]
```diff
@@ -319,14 +326,27 @@ func (b *Builder) buildStore(
 	expectedType interface{},
 	listWatchFunc func(kubeClient clientset.Interface, ns string) cache.ListerWatcher,
 ) cache.Store {
-	filteredMetricFamilies := generator.FilterMetricFamilies(b.whiteBlackList, metricFamilies)
+	filteredMetricFamilies := generator.FilterMetricFamilies(b.whiteBlackList, b.enableAggregations, metricFamilies)
```
Comment (marktwallace) [inline internal/store/builder.go:329]: It's hard to follow all the dependencies unless I pull the branch, so I'll just ask here: does FilterMetricFamilies take into account the "--metric-denylist" ?

File: internal/store/certificatesigningrequest.go [hunk 2]
```diff
@@ -79,6 +80,9 @@ var (
 					Metrics: addCSRConditionMetrics(csr.Status),
 				}
 			}),
+			AggregateBy: map[string]aggregation.Aggregation{
+				"condition": aggregation.ByLabels("condition"),
+			},
```
Comment (marktwallace) [inline internal/store/certificatesigningrequest.go:85]: I like this construct -- it is elegant to just  include an AggregateBy section in the FamilyGenerator

File: internal/store/endpoint.go [hunk 3]
```diff
@@ -102,6 +103,9 @@ var (
 					},
 				}
 			}),
+			AggregateBy: map[string]aggregation.Aggregation{
+				"namespace": aggregation.ByLabels("namespace"),
```
Comment (marktwallace) [inline internal/store/endpoint.go:107]: This is just a thought, that maybe some of the generality that you have here in the "ByLabels" clause, that it might be possible to make that accessible through the startup arguments for KSM. (We would not need that for internal purposes, but it might be helpful for an OSS PR.)

File: internal/store/hpa.go [hunk 4]
```diff
@@ -223,6 +224,9 @@ var (
 					Metrics: ms,
 				}
 			}),
+			AggregateBy: map[string]aggregation.Aggregation{
+				"namespace": aggregation.ByLabels("namespace", "condition", "status"),
```
Comment (marktwallace) [inline internal/store/hpa.go:228]: I'm a little confused here as to what it means to add all three -- "namespace", "condition", "status" -- does that generate three separate aggregates? In the .md file above you only call out kube_hpa_status_condition_sum_by_namespace.

File: internal/store/job.go [hunk 5]
```diff
@@ -152,6 +153,10 @@ var (
 					},
 				}
 			}),
+			AggregateBy: map[string]aggregation.Aggregation{
+				"namespace": aggregation.ByLabels("namespace"),
+				"owner": aggregation.ByNamespaceAndOwner(),
```
Comment (marktwallace) [inline internal/store/job.go:158]: OK, I guess that "owner" is not there as a label, so some generic code is required to sum by owner, right?

File: pkg/metric_aggregation/aggregation.go [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+	AggregationFunc func(interface{}, metric.FamilyInterface) []AggregatedValue
+}
+
+// AggregationSet represents all aggregations available from a single FamilyGenerator
+type AggregationSet struct {
+	aggregations map[string]*AggregatedMetric
+
+	// Aggregations are calculated lazily: an update to the stored metrics
+	// sets this flag to false and then the aggregates are recalculated
+	// during WriteAll().
```
Comment (marktwallace) [inline pkg/metric_aggregation/aggregation.go:54]: This is interesting -- the overhead for the aggregation calculation will be part of the latency of the polling request from the prometheus server, right? I don't know what other overhead is normally built into the handling of the /metrics request, but I guess in some extreme cases the calculation of aggregates could be expensive. Since the AggregationFunc is (apparently) a generic interface, we will need to be careful with what functions are allowed, to make sure they would not slow down metrics collection for the prom server. Is my reasoning correct here?

File: pkg/metric_aggregation/aggregation.go [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+func (a *AggregatedMetric) GetMetricFamily(shard int) metric.Family {
+	aggValue := make(map[string]float64)
+	labelValueSets := make(map[string][]string)
+	for _, obj := range a.Objects {
+		for _, m := range obj {
+			labelString := fmt.Sprintf("%#v", m.LabelValues)
+			labelValueSets[labelString] = m.LabelValues
+
+			// this is what makes all the aggregations sums
+			aggValue[labelString] += m.Value
```
Comment (marktwallace) [inline pkg/metric_aggregation/aggregation.go:132]: So at the moment, summing is the only aggregation, right? So not very expensive given that all the floats are in memory.

File: pkg/metric_aggregation/aggregators.go [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+			owners := meta.GetOwnerReferences()
+			if len(owners) == 0 {
+				labelSets = [][]string{{ns, "<none>", "<none>", "<none>"}}
+			} else {
+				labelSets = make([][]string, len(owners))
+
+				for i, owner := range owners {
+					if owner.Controller != nil {
+						labelSets[i] = []string{ns, owner.Kind, owner.Name, strconv.FormatBool(*owner.Controller)}
```
Comment (marktwallace) [inline pkg/metric_aggregation/aggregators.go:76]: This is just a note for myself to go back and look at this after I have pulled the branch. Sorry for being so chatty in the comments :-). (Actually I *really* like PR comments myself.)

File: internal/store/node.go [hunk 9]
```diff
@@ -532,3 +551,44 @@ func createNodeListWatch(kubeClient clientset.Interface, ns string) cache.Lister
 		},
 	}
 }
+
+func byRole(labels ...string) aggregation.Aggregation {
```
Comment (marktwallace) [inline internal/store/node.go:555]: This might be really nit-picky or wrong, but in the "AggregateBy:" declarations, "ByLabels" starts with upper case, but "byRole" is lowercase. Shouldn't byRole have public visibility to be used in the "AggregateBy:" ?

I'll admit, even reading the code in context after I pulled the path, I do not fully understand the implementation of byRole -- I might need to grok the whole pattern of internal/store in order to fully follow the code.

I can't figure out why byRole is in node.go and the other functions like ByLabels are in aggregators.go -- I would have expected to see them all together.

File: pkg/metric_aggregation/aggregators.go (hunk 8 above)
Comment (marktwallace) [reply pkg/metric_aggregation/aggregators.go:76]: Ok, now that I read it in context it all makes sense.

**Review by marktwallace — APPROVED** [review]
OK, I'm all done. I had a lot of questions, but I did not see anything that would prevent me from approving the PR. Most of my questions were about the overall way it works. I liked the coding patterns, and think they are consistent with the rest of Kube-state-metrics. I think this is in good shape for our internal use, but I would want to take another pass before sending the PR to the K8s project. All that said, I pulled the path, but I did not run the tests, etc. 

Great work, and thanks for showing it to me!

File: internal/store/builder.go (hunk 1 above)
Comment (gnosek) [reply internal/store/builder.go:329]: Yes, it's the whole purpose of the function. Without aggregations, the only two outcomes are "metric remains in the list" and "metric is removed from the list". With aggregations, there's a new third state: the un-aggregated metric is filtered out, only (some or all) aggregations remain. This requires the metric to remain in the list after filtering but we set a flag to only send the aggregations.

File: internal/store/endpoint.go (hunk 3 above)
Comment (gnosek) [reply internal/store/endpoint.go:107]: I'd leave this for the OSS community to decide. The feature would certainly be nice but I'm worried about the interface (would this be usable without an extremely long command line).

File: internal/store/hpa.go (hunk 4 above)
Comment (gnosek) [reply internal/store/hpa.go:228]: No, this creates a single aggregation called `namespace` (i.e. `kube_hpa_status_condition_sum_by_namespace`) aggregated by the three labels, so you'll end up with e.g.

```
kube_hpa_status_condition_sum_by_namespace{namespace="X1",condition="Y1",status="Z1"} 1
kube_hpa_status_condition_sum_by_namespace{namespace="X1",condition="Y1",status="Z2"} 2
kube_hpa_status_condition_sum_by_namespace{namespace="X1",condition="Y2",status="Z1"} 3
kube_hpa_status_condition_sum_by_namespace{namespace="X1",condition="Y2",status="Z2"} 5
```

etc. Otherwise, we'd just sum all the HPAs per namespace and lose the status information.

File: internal/store/job.go (hunk 5 above)
Comment (gnosek) [reply internal/store/job.go:158]: Yes, exactly. This was the reason I added aggregations by arbitrary functions in the first place but I kind of like how it turned out in the end.

File: internal/store/node.go (hunk 9 above)
Comment (gnosek) [reply internal/store/node.go:555]: `byRole` doesn't need to be public as we only refer to it in the same file. I put it here because it really is node-specific.

The logic is:
- take all the labels starting with `node-role.kubernetes.io/` (though we expect just one, there technically may be more)
- strip the prefix
- if there are no such labels, use `<none>` as a fallback

For each role (again, only one is really expected), aggregate the metric into each individual role.

The aggregation functions are expected to return effectively an array of (metric value, labels) pairs. That array (combined across all the relevant k8s objects) is then aggregated by adding together all the values with matching labels. This means an individual metric can be a part of multiple aggregated metrics, even by the same criteria (it would be weird but it's technically valid and a nasty edge case to handle otherwise).

File: pkg/metric_aggregation/aggregation.go (hunk 6 above)
Comment (gnosek) [reply pkg/metric_aggregation/aggregation.go:54]: The aggregation functions are called in response to cluster state updates, not during the `/metrics` request. The lazy part is adding all the values together.

File: pkg/metric_aggregation/aggregation.go (hunk 7 above)
Comment (gnosek) [reply pkg/metric_aggregation/aggregation.go:132]: Yes. Memory is the bigger cost (an estimate is ~3x the size of the whole `/metrics` response in extra memory usage). It actually might be better _not_ to cache the aggregation result and add up the numbers every time (trading CPU time and a bit of latency for reduced memory usage).

**Comment by gnosek** [comment]
> (1) These look really useful -- especially the sum_by_owner aggregates.

Thanks :)

> (2) There are a lot of new metrics here.
> (3) If we send a PR to the OSS project, we might get push-back on the overall number of metrics. There could also be multiple points of view as to whether these should be aggregated at this level -- it is good for our use case, but users that do not have the same problems of scale and data retention may think that using labels to aggregate with PromQL is better, and this would look redundant to them.

Yup. I went through the metrics to see what could be useful, didn't really take into account our internal needs so we'll probably want to revisit it.

The aggregations are disabled by default, so if you don't need them, you don't pay the cost. In a perfect world, I'd use PromQL too but we'd be dead long before we put the metrics anywhere that supports PromQL.

> (4) Since our flag "--enable-aggregations" is orthogonal to to the "--metric-allowlist" and "--metric-denylist" flags, it would bring in aggregates for things that are denied, right? That could be a source of confusion for OSS users.

If you `--enable-aggregations --metric-denylist "^kube_pod_phase$"` then you will get the `kube_pod_phase_sum_by_...` metrics but with `--enable-aggregations --metric-denylist "^kube_pod_phase"` (or just `--enable-aggregations --metric-denylist kube_pod_phase`) you won't. Aggregations respect the filtering rules just like normal metrics.

> I'll keep adding comments as I go. I'm liking the code so far :-)

Thanks :)
